hypervisor/dm/vpci/vpci.c

```diff
@@ -370,30 +370,101 @@ static void vpci_deinit_pt_dev(struct pci_vdev *vdev)
 	deinit_vmsi(vdev);
 }
 
-static int32_t write_pt_dev_cfg(struct pci_vdev *vdev, uint32_t offset,
-		uint32_t bytes, uint32_t val)
+struct cfg_header_perm {
+	/* per bit for per 4 Bytes registers, only low 16 bit is used.
+	 * for example, bit 0 for CFG Vendor ID and Device ID register,
+	 * Bit 1 for CFG register Command and Status register and so on
+	 */
+
+	/* When one bit set, it means its corresponding 4 Bytes registers is pass through,
+	 * Otherwise, it's virtualized.
+	 */
+	uint32_t pt_mask;
+
+	/* When one bit set, it means its corresponding 4 Bytes registers is read-only,
+	 * Otherwise, it's writable.
+	 */
+	uint32_t ro_mask;
+};
+
+static const struct cfg_header_perm cfg_hdr_perm = {
+	/* Only Command (0x04-0x05) and Status (0x06-0x07) Registers are pass through */
+	.pt_mask = 0x0002U,
+	/* Command (0x04-0x05) and Status (0x06-0x07) Registers and
+	 * Base Address Registers (0x10-0x27) are writable */
+	.ro_mask = (uint16_t)~0x03f2U
+};
+
+
+/*
+ * @pre offset + bytes < PCI_CFG_HEADER_LENGTH
+ */
+static void read_cfg_header(const struct pci_vdev *vdev,
+		uint32_t offset, uint32_t bytes, uint32_t *val)
+{
+	if (vbar_access(vdev, offset)) {
+		/* bar access must be 4 bytes and offset must also be 4 bytes aligned */
+		if ((bytes == 4U) && ((offset & 0x3U) == 0U)) {
+			*val = pci_vdev_read_bar(vdev, pci_bar_index(offset));
+		} else {
+			*val = ~0U;
+		}
+	} else {
+		if (bitmap32_test((uint16_t)offset >> 2U, &cfg_hdr_perm.pt_mask)) {
+			*val = pci_pdev_read_cfg(vdev->pdev->bdf, offset, bytes);
+		} else {
+			*val = pci_vdev_read_cfg(vdev, offset, bytes);
+		}
+	}
+}
+
+/*
+ * @pre offset + bytes < PCI_CFG_HEADER_LENGTH
+ */
+static void write_cfg_header(struct pci_vdev *vdev,
+		uint32_t offset, uint32_t bytes, uint32_t val)
 {
 	if (vbar_access(vdev, offset)) {
 		/* bar write access must be 4 bytes and offset must also be 4 bytes aligned */
 		if ((bytes == 4U) && ((offset & 0x3U) == 0U)) {
 			vdev_pt_write_vbar(vdev, pci_bar_index(offset), val);
 		}
+	} else {
+		if (offset == PCIR_COMMAND) {
+#define PCIM_SPACE_EN (PCIM_CMD_PORTEN | PCIM_CMD_MEMEN)
+			uint16_t phys_cmd = (uint16_t)pci_pdev_read_cfg(vdev->pdev->bdf, PCIR_COMMAND, 2U);
+
+			/* check whether need to restore BAR because some kind of reset */
+			if (((phys_cmd & PCIM_SPACE_EN) == 0U) && ((val & PCIM_SPACE_EN) != 0U) &&
+					pdev_need_bar_restore(vdev->pdev)) {
+				pdev_restore_bar(vdev->pdev);
+			}
+		}
+
+		if (!bitmap32_test((uint16_t)offset >> 2U, &cfg_hdr_perm.ro_mask)) {
+			if (bitmap32_test((uint16_t)offset >> 2U, &cfg_hdr_perm.pt_mask)) {
```

Comment:
Not sure on this kind of type casting

Reply:
I will check will huihuang about this,

Reply:
Add brackets between (uint16_t)offset to eliminate the ambiguity between casting  and the operator ">>". The casting here is needed.